include/sound/sof/ipc4/header.h

```diff
@@ -0,0 +1,338 @@
+	SOF_IPC4_GLB_LOAD_LIBRARY,
+	SOF_IPC4_GLB_INTERNAL_MESSAGE = 26,
+
+	/* Notification (FW to SW driver) */
+	SOF_IPC4_GLB_NOTIFICATION,
+	SOF_IPC4_GLB_MAX_IXC_MESSAGE_TYPE = 31
+};
+
+/* Message direction */
+/* The maximum size of the enum is 1 as the enum must fit into 1 bit*/
+enum sof_ipc4_msg_dir {
+	SOF_IPC4_MSG_REQUEST = 0,
+	SOF_IPC4_MSG_REPLY,
+};
+
+enum sof_ipc4_pipe_line_state {
+	SOF_IPC4_PIPE_INVALID_STATE = 0,
+	SOF_IPC4_PIPE_UNINITIALIZED,
+	SOF_IPC4_PIPE_RESET,
+	SOF_IPC4_PIPE_PAUSED,
+	SOF_IPC4_PIPE_RUNNING,
+	SOF_IPC4_PIPE_EOS
+};
+
+/* global common ipc msg */
+#define SOF_IPC4_GLB_MSG_TARGET_SHIFT		30
+#define SOF_IPC4_GLB_MSG_TARGET_MASK		BIT(30)
+#define SOF_IPC4_GLB_MSG_TARGET(x)		((x) << SOF_IPC4_GLB_MSG_TARGET_SHIFT)
+
+#define SOF_IPC4_GLB_MSG_DIR_SHIFT		29
+#define SOF_IPC4_GLB_MSG_DIR_MASK		BIT(29)
+#define SOF_IPC4_GLB_MSG_DIR(x)			((x) << SOF_IPC4_GLB_MSG_DIR_SHIFT)
+
+#define SOF_IPC4_GLB_MSG_TYPE_SHIFT		24
+#define SOF_IPC4_GLB_MSG_TYPE_MASK		GENMASK(28, 24)
+#define SOF_IPC4_GLB_MSG_TYPE(x)		((x) << SOF_IPC4_GLB_MSG_TYPE_SHIFT)
+
+/* pipeline creation ipc msg */
+#define SOF_IPC4_GLB_PIPE_INSTANCE_SHIFT	16
+#define SOF_IPC4_GLB_PIPE_INSTANCE_MASK		GENMASK(23, 16)
+#define SOF_IPC4_GLB_PIPE_INSTANCE_ID(x)	((x) << SOF_IPC4_GLB_PIPE_INSTANCE_SHIFT)
+
+#define SOF_IPC4_GLB_PIPE_PRIORITY_SHIFT	11
+#define SOF_IPC4_GLB_PIPE_PRIORITY_MASK		GENMASK(15, 11)
+#define SOF_IPC4_GLB_PIPE_PRIORITY(x)		((x) << SOF_IPC4_GLB_PIPE_PRIORITY_SHIFT)
+
+#define SOF_IPC4_GLB_PIPE_MEM_SIZE_SHIFT	0
+#define SOF_IPC4_GLB_PIPE_MEM_SIZE_MASK		GENMASK(10, 0)
+#define SOF_IPC4_GLB_PIPE_MEM_SIZE(x)		((x) << SOF_IPC4_GLB_PIPE_MEM_SIZE_SHIFT)
+
+#define SOF_IPC4_GL_PIPE_EXT_LP_SHIFT		0
+#define SOF_IPC4_GL_PIPE_EXT_LP_MASK		BIT(0)
+#define SOF_IPC4_GL_PIPE_EXT_LP(x)		((x) << SOF_IPC4_GL_PIPE_EXT_LP_SHIFT)
+
+/* pipeline set state ipc msg */
+#define SOF_IPC4_GL_PIPE_STATE_TYPE_SHIFT	24
+#define SOF_IPC4_GL_PIPE_STATE_TYPE_MASK	GENMASK(28, 24)
+#define SOF_IPC4_GL_PIPE_STATE_TYPE(x)		((x) << SOF_IPC4_GL_PIPE_STATE_TYPE_SHIFT)
+
+#define SOF_IPC4_GL_PIPE_STATE_ID_SHIFT		16
+#define SOF_IPC4_GL_PIPE_STATE_ID_MASK		GENMASK(23, 16)
+#define SOF_IPC4_GL_PIPE_STATE_ID(x)		((x) << SOF_IPC4_GL_PIPE_STATE_ID_SHIFT)
+
+#define SOF_IPC4_GL_PIPE_STATE_SHIFT		0
+#define SOF_IPC4_GL_PIPE_STATE_MASK		GENMASK(15, 0)
+#define SOF_IPC4_GL_PIPE_STATE(x)		((x) << SOF_IPC4_GL_PIPE_STATE_SHIFT)
+
+enum sof_ipc4_sampling_frequency {
+	SOF_IPC4_FS_8000HZ	= 8000,
+	SOF_IPC4_FS_11025HZ	= 11025,
+	SOF_IPC4_FS_12000HZ	= 12000,
+	SOF_IPC4_FS_16000HZ	= 16000,
+	SOF_IPC4_FS_18900HZ	= 18900,
+	SOF_IPC4_FS_22050HZ	= 22050,
+	SOF_IPC4_FS_24000HZ	= 24000,
+	SOF_IPC4_FS_32000HZ	= 32000,
+	SOF_IPC4_FS_37800HZ	= 37800,
+	SOF_IPC4_FS_44100HZ	= 44100,
+	SOF_IPC4_FS_48000HZ	= 48000, /* Default. */
+	SOF_IPC4_FS_64000HZ	= 64000,
+	SOF_IPC4_FS_88200HZ	= 88200,
+	SOF_IPC4_FS_96000HZ	= 96000,
+	SOF_IPC4_FS_176400HZ	= 176400,
+	SOF_IPC4_FS_192000HZ	= 192000,
```

Comment:
There are existing boards with 384KHz (!) audio support via ALC4082 ;)
Why this needs to be an `enum`?
If this is passed within an IPC message then the important thing is the number of bits used to store this information...

Reply:
384k ?  " The AK4499EQ supports PCM and DSD512 inputs up to 768kHz".  My idea is we can add it for future, but now we have to align to audio sdk

Reply:
I agree that we don't really need the enum. The purpose of using the enum is to avoid using the frequency value directly.
Given that the code is merged on the firmware side, I prefer to keep it as it is. And we can send follow up patches to extend the enum on both kernel and firmware side.
https://github.com/thesofproject/sof/blob/main/src/include/ipc4/base-config.h

Reply:
My point is that the rate is represented in Linux as `unsigned int` as _number_, so you are going to have something like:
```c
enum sof_ipc4_sampling_frequency fs = params_rate(params);
```
and in this case I don't think enum on the left side can protect you from a rate which is not listed under the `enum sof_ipc4_sampling_frequency`
This works like charm on a quick test:
```c
    enum sof_ipc4_sampling_frequency fs;
    unsigned int rate = 63;
    int i;

    for (i = 0; i < 3; i++)
	rate *= 10;

    printf("rate: %u\n", rate);
    fs = rate;
    printf("fs: %u\n", fs);
    fs = 62000;
    printf("fs2: %u\n", fs);
```
```
rate: 63000
fs: 63000
fs2: 63000
```
Even sparse have no complains about this...

Reply:
Hmm, true @ujfalusi . Given the enum value matches the actual rate, why not just leave this out? The enum can be defined in implementation (maybe to do a is_valid(rate) check), but it should not be part of IPC header if the interface is to pass the rate as a number.